plutus-core/plutus-core/src/PlutusCore/TypeCheck/Internal.hs

```diff
@@ -577,10 +578,28 @@ inferTypeM (Case ann resTy scrut branches) = do
             -- for the number of branches
             Nothing -> throwError (TypeMismatch ann (void scrut) expectedSop vScrutTy)
         vTy -> case annotateCaseBuiltin vTy branches of
-            Right branchesAndArgTypes -> for_ branchesAndArgTypes $ \(c, argTypes) -> do
+            Right branchesAndArgTypes -> for_ branchesAndArgTypes $ \case
+              FixedArityBranch c argTypes -> do
                 vArgTypes <- traverse (fmap unNormalized . normalizeTypeM) argTypes
                 -- made of sub-parts of a normalized type, so normalized
                 checkTypeM ann c (Normalized $ mkIterTyFun () vArgTypes (unNormalized vResTy))
+              VariableArityBranch c argType -> do
```

Comment:
Not really happy here, but don't think there's any better way to go about without bigger changes. The fact that each branch can have arbitrary number of arguments introduces this dynamic behavior. So, it is only possible to check the return type and to check if are inputs(of arbitrary number) have correct type.